### DISCUSSION TITLE
Is dynamic schema interesting for pydantic ?

### DISCUSSION THREAD

**Discussion by sapetnioc** [pydantic/pydantic#3227] [Question]
I would like to have the possibility to have a dynamic schema allowing to add and remove fields after class creation. Is it of any interest for Pydantic project ? Is it worth to make an API change proposal in a ticket (this would be the first step before a pull request) ?

A very rough idea of what this change could be is to add a `DynamicBaseModel` class that would add `add_field()` and `remove_field()` methods on models.

I made some proof of concept code and, at first sight, it doesn't seem too complex to implement by storing dynamic fields information (e.g. `__fields__`, `__fields_set__`) on the instance instead of the class.

**Comment by samuelcolvin** [comment]
You can already use `create_model` with a base class. I don't think we need anything else.

**Reply by sapetnioc** [reply]
Well, with `create_model` you get a class that does not allow to add or remove fields on the fly. The only possibility would be to copy the whole model on each single field change which is not efficient. I believe that this dynamic field feature should be in Pydantic because its implementation required to play with a few model internals fields (at to change some methods from static to instance).

My goal is to implement a class where some fields could be modified according to other fields values. And to have a graphical interface for that (with a modification notification system). We already have an implementation based on traits but Pydantic seem to be a better base (lighter an more Pythonic). I think the whole thing is beyond the scope of Pydantic but, in my opinion, dynamic fields could be a plus to the project if enough people are interested by that feature.
